massa-execution-worker/src/context.rs

```diff
@@ -635,6 +635,54 @@ impl ExecutionContext {
         )
     }
 
+    /// Try to slash 1 roll from the denounced address.
+    ///
+    /// # Arguments
+    /// * `denounced_addr`: address to slash the roll or deferred credit from
+    pub fn try_slash_roll(
+        &mut self,
+        denounced_addr: &Address,
+        roll_count: u64,
+    ) -> Result<Amount, ExecutionError>
+    {
+        let slashed_coins;
+
+        // First try to slash roll if any
+        if self.speculative_roll_state.try_burn_rolls(
+            denounced_addr,
+            roll_count,
+        ).is_err() {
+
+            // if no roll were slashed, that can likely be done from deferred credit
+            let credits = self.speculative_roll_state.get_address_deferred_credits(denounced_addr, self.slot);
+            let amount_to_slash = self.config.roll_price
+                .checked_mul_u64(roll_count)
+                .ok_or_else(||
+                    ExecutionError::RuntimeError(format!("Cannot mult roll price by {}", roll_count))
+                )?;
+
+            // Remove deferred credits starting at self.slot until
+            let mut amount_to_rm = amount_to_slash;
+            for (slot, _amount) in credits.iter() {
+                if amount_to_rm == Amount::zero() {
+                    break;
+                }
+                let amount_slashed_at_slot = self.speculative_roll_state
+                    .remove_deferred_credits(slot, denounced_addr, &amount_to_rm);
+                amount_to_rm = amount_to_rm.saturating_sub(amount_slashed_at_slot)
```

Comment:
I feel like the whole loop iterating on credits is not useful. I would do it like this:
1- try to slash the rolls, list the number of coins that we failed to slash
2- just call remove_deferred_credits (rename it slash_deferred_credits) on the remainder, if any

Reply:
As we discuss it, remove_deferred_credits slash a given amount at a given slot so we need the loop here ;)